[Dev] Add player IV override and player + enemy Nature override

DayKev opened the pull request:
## What are the changes the user will see?
N/A

## Why am I making these changes?
Improved testing (both manual and automatic).

## What are the changes from a developer perspective?
- New overrides added: `IVS_OVERRIDE`, `NATURE_OVERRIDE` and `ENEMY_NATURE_OVERRIDE`.
- Helper functions added to `OverridesHelper`.
- All tests now automatically set every Pokémon's nature to a neutral nature and all IVs to 31 by default. This is disabled whenever any of the override helpers are used.

## Screenshots/Videos
<details><summary>Example</summary>
<p>

```ts
const overrides = {
  NATURE_OVERRIDE: Nature.ADAMANT,
  ENEMY_NATURE_OVERRIDE: Nature.BOLD,
  IVS_OVERRIDE: 12,
  ENEMY_IVS_OVERRIDE: [1, 2, 3, 4, 5, 6],
} satisfies Partial<InstanceType<typeof DefaultOverrides>>;
```

![image](https://github.com/user-attachments/assets/12f41beb-99cd-4128-910d-bb33619d5020)

</p>
</details> 

## Checklist
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test:silent`)
  - [x] Have I created new automated tests (`npm run test:create`) or updated existing tests related to the PR's changes?

DayKev committed:
Add player IV override and player + enemy nature overrides

DayKev committed:
Add new overrides to the `OverridesHelper` class

DayKev committed:
Apply IV and Nature overrides automatically to all tests

DayKev committed:
Fix Wimp Out tests

DayKev committed:
Add TSDocs to new overrides

flx-sta commented on test/test-utils/helpers/overrides-helper.ts:
```diff
@@ -407,13 +425,76 @@ export class OverridesHelper extends GameManagerHelper {
+    if (nature === null) {
+      this.log("Player Nature override disabled!");
+    } else {
+      this.log(`Player Nature set to ${Nature[nature]} (${nature})!`);
```
```suggestion
      this.log(`Player Nature set to ${Nature[nature]} (=${nature})!`);
```

flx-sta commented on test/test-utils/helpers/overrides-helper.ts:
```diff
@@ -407,13 +425,76 @@ export class OverridesHelper extends GameManagerHelper {
+    if (nature === null) {
+      this.log("Enemy Nature override disabled!");
+    } else {
+      this.log(`Enemy Nature set to ${Nature[nature]} (${nature})!`);
```
```suggestion
      this.log(`Enemy Nature set to ${Nature[nature]} (=${nature})!`);
```

Tempo-anon commented on src/battle-scene.ts:
```diff
@@ -977,14 +999,25 @@ export default class BattleScene extends SceneBase {
-      ENEMY_IVS_OVERRIDE_VALIDATED = Overrides.ENEMY_IVS_OVERRIDE;
-    } else if (typeof Overrides.ENEMY_IVS_OVERRIDE === "number") {
-      ENEMY_IVS_OVERRIDE_VALIDATED = new Array(6).fill(Overrides.ENEMY_IVS_OVERRIDE);
+    if (Overrides.ENEMY_IVS_OVERRIDE === null) {
```
The IV check validation for it being a number or array of numbers with a proper length and value between 0 and 31 is duplicated

flx-sta replied:
@Tempo-anon Where is it duplicate?

Tempo-anon replied:
Lines 940-960

flx-sta replied:
I see. I've added code suggestions to clarify

Tempo-anon commented on test/test-utils/helpers/overrides-helper.ts:
```diff
@@ -32,14 +34,30 @@ import { expect, vi } from "vitest";
+   * If `true`, will set the IV overrides for player and enemy pokemon to `31` at the start of each test.
+   * @defaultValue `true`
+   */
+  public normalizeIVs: boolean = true;
```
neat! having all tests by default forcing natures and IVs will be handy for preventing flaky discrepencies

Tempo-anon commented on src/battle-scene.ts:
```diff
@@ -977,14 +999,25 @@ export default class BattleScene extends SceneBase {
+      // do nothing
+    } else if (Array.isArray(Overrides.ENEMY_IVS_OVERRIDE)) {
+      if (Overrides.ENEMY_IVS_OVERRIDE.length !== 6) {
+        throw new Error("The Player IVs override must be an array of length 6 or a number!");
```
The error message is for player IV despite this being enemy IV overriding

flx-sta replied:
```suggestion
        throw new Error("The Enemy IVs override must be an array of length 6 or a number!");
```

flx-sta commented on src/battle-scene.ts:
```diff
@@ -977,14 +999,25 @@ export default class BattleScene extends SceneBase {
+        throw new Error("The Player IVs override must be an array of length 6 or a number!");
+      }
+      if (Overrides.ENEMY_IVS_OVERRIDE.some((value) => !isBetween(value, IV_MIN, IV_MAX))) {
+        throw new Error(`All IVs in the player IV override must be between ${IV_MIN} and ${IV_MAX}!`);
```
```suggestion
        throw new Error(`All IVs in the enemy IV override must be between ${IV_MIN} and ${IV_MAX}!`);
```

flx-sta commented on src/battle-scene.ts:
```diff
@@ -977,14 +999,25 @@ export default class BattleScene extends SceneBase {
+      pokemon.ivs = Overrides.ENEMY_IVS_OVERRIDE;
+    } else {
+      if (!isBetween(Overrides.ENEMY_IVS_OVERRIDE, IV_MIN, IV_MAX)) {
+        throw new Error(`The Player IV override must be a value between ${IV_MIN} and ${IV_MAX}!`);
```
```suggestion
        throw new Error(`The Enemy IV override must be a value between ${IV_MIN} and ${IV_MAX}!`);
```

flx-sta committed:
Apply suggestions from code review

- fix console messages for enemy IV override
- add `=` prefix in override helper
